pcmanfm/mainwindow.cpp

```diff
@@ -1256,6 +1267,80 @@ void MainWindow::focusPathEntry() {
     }
 }
 
+void MainWindow::dragEnterEvent(QDragEnterEvent* event) {
+    if(event->mimeData()->hasFormat("application/pcmanfm-qt-tab")) {
+        event->acceptProposedAction();
+    }
+}
+
+void MainWindow::dropEvent(QDropEvent* event) {
+    if(event->mimeData()->hasFormat("application/pcmanfm-qt-tab")) {
+        dropTab();
+    }
+    event->acceptProposedAction();
+}
+
+void MainWindow::dropTab() {
+    if(lastActive_ == nullptr // impossible
+       || lastActive_ == this) { // don't drop on the same window
+        ui.tabBar->finishMouseMoveEvent();
+        return;
+    }
+
+    // close the tab in the first window and add
+    // its page to a new tab in the second window
+    TabPage* dropPage = lastActive_->currentPage();
+    if(dropPage) {
+        disconnect(dropPage, &TabPage::titleChanged, lastActive_, &MainWindow::onTabPageTitleChanged);
+        disconnect(dropPage, &TabPage::statusChanged, lastActive_, &MainWindow::onTabPageStatusChanged);
+        disconnect(dropPage, &TabPage::openDirRequested, lastActive_, &MainWindow::onTabPageOpenDirRequested);
+        disconnect(dropPage, &TabPage::sortFilterChanged, lastActive_, &MainWindow::onTabPageSortFilterChanged);
+        disconnect(dropPage, &TabPage::backwardRequested, lastActive_, &MainWindow::on_actionGoBack_triggered);
+        disconnect(dropPage, &TabPage::forwardRequested, lastActive_, &MainWindow::on_actionGoForward_triggered);
+
+        // release mouse before tab removal because otherwise, the source tabbar
+        // might not be updated properly with tab reordering during a fast drag-and-drop
+        lastActive_->ui.tabBar->releaseMouse();
+
+        QWidget* page = lastActive_->ui.stackedWidget->currentWidget();
+        lastActive_->ui.stackedWidget->removeWidget(page);
+        int index = addTabWithPage(dropPage);
+        ui.tabBar->setCurrentIndex(index);
+    }
+    else {
+        ui.tabBar->finishMouseMoveEvent(); // impossible
+    }
+}
+
+void MainWindow::detachTab() {
+    if (ui.stackedWidget->count() == 1) { // don't detach a single tab
+        ui.tabBar->finishMouseMoveEvent();
+        return;
+    }
+
+    // close the tab and move its page to a new window
+    TabPage* dropPage = currentPage();
+    if(dropPage) {
+        disconnect(dropPage, &TabPage::titleChanged, this, &MainWindow::onTabPageTitleChanged);
```

Comment:
Are you disconnecting all signals here? If so, you can use on of the `disconnect`'s overload to do it with one call.

Reply:
Just for the sake of caution: one-by-one disconnecting means we exactly know what's happening.

However, feel free to do it at once by adding a commit for both cases here.

Reply:
@palinek  Never mind! On second thought, that "caution" seems too much here. Will change it later, today.

Long story short, I had to disconnect signals in FeatherPad one by one and did the same thing here without it being really needed